Improve template

elliotcm opened the pull request:
This upgrades and/or adds functionality in the generated app.

Best reviewed commit by commit.  Example of the output here: https://github.com/alphagov/govuk-developer-tools/tree/f4083c9cfcd78b4b4186301eeb649b2f147ecefa

elliotcm committed:
Upgrade to ruby 2.3.0

elliotcm committed:
README whitespace & tweaks.

elliotcm committed:
Add useful testing tools.

Adds webmock, timecop, and factory girl.

elliotcm committed:
Add GDS SSO.

elliotcm committed:
Add standardised `rails_` and `spec_helper`s.

elliotcm committed:
Add common debuggers.

elliotcm committed:
Add deprecated_columns.

Assumes we're going to have a DB for at least
GDS SSO.  Helps when removing columns from tables.

elliotcm committed:
Make rcov/simplecov optional but enabled for jenkins.

It can misbehave on certain systems and we don't
need it in dev.

elliotcm committed:
Upgrade some gems.

elliotcm committed:
Start with a lean Gemfile.

elliotcm committed:
Clean up Gemfile modification.

elliotcm committed:
Standardise on double quotes.

tijmenb commented:
Coolio.

Is there a way to pass arguments to the generator? These changes are mostly relevant for backend apps, and probably not suitable for frontends. Or perhaps we make it explicit that we only generate backend apps with this.

elliotcm commented:
@tijmenb mm.  If you're happy with these changes in principle, merge it and I'll split out / parameterize in another PR.

elliotcm commented:
@tijmenb actually nevermind, I'll do it as part of this PR

elliotcm committed:
Restructure templates for each type of app we have.

See the classes in `lib` to see what exactly each
class of app sets up.

elliotcm commented:
@tijmenb Updated.

elliotcm commented:
Links to https://github.com/alphagov/styleguides/pull/90

boffbowsh commented on lib/govuk_publishing_template.rb:
```diff
@@ -0,0 +1,8 @@
+
+class GovukPublishingTemplate < GovukAdminTemplate
+  def apply
+    add_sidekiq
```
Shouldn't this call `super` first?

elliotcm replied:
Yes, it should, thanks.  I've updated the commit and tested manually.

I've also added two more commits so if you could check those that'd be great.  Otherwise good to merge.

boffbowsh commented:
One comment, but this looks 👏

elliotcm committed:
Clarify post-build instructions.

Add a header to make it clear that these are things
that need doing rather than things that have been
done.

Also adds --skip-bundle to the invocation as this
last bundle run is unnecessary and pushed the post-
build instructions off the screen.

elliotcm committed:
Add a script to invoke rails new.

Does some sense checks on app name and Rails version.

tijmenb commented on bin/install.sh:
```diff
@@ -0,0 +1,28 @@
+}
+
+[ "$#" -eq 2 ] || die "2 arguments required, $# provided"
+echo $1 | grep -E -q '^[a-z-]+$' || die "Application name must be lowercase with hyphens"
```
❤️

tijmenb commented:
This turned out rally great 👍 👍 👍
